### PR TITLE
feat(auth) implementation of SAML

### PR DESCRIPTION
**Pull request by wtermellil**
## Description

Implementation of SAML

https://user-images.githubusercontent.com/110023866/225228267-cb26097a-8e98-454a-a54a-c2354d736c9f.mp4

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [x] 23.04.x (master)

<h2> How this pull request can be tested ? </h2>

Configure your SAML provider and you need to authenticate

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Security/Authentication/Infrastructure/Api/Login/SAML/CallbackController.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        switch (true) {
+            case is_a($presenter->getResponseStatus(), PasswordExpiredResponse::class)
+                || is_a($presenter->getResponseStatus(), UnauthorizedResponse::class)
+                || is_a($presenter->getResponseStatus(), ErrorResponse::class):
+                return View::createRedirect(
+                    $this->getBaseUrl() . '/login?authenticationError=' . $presenter->getResponseStatus()->getMessage()
+                );
+            case is_a($presenter->getResponseStatus(), ErrorAclConditionsResponse::class):
+            case is_a($presenter->getResponseStatus(), ErrorAuthenticationConditionsResponse::class):
```
Comment (abuathier) [inline centreon/src/Core/Security/Authentication/Infrastructure/Api/Login/SAML/CallbackController.php:75]: `instanceof` is far better than `is_a` when using object instances (`->getResponseStatus()`).
- more expressive
- faster (at least x2)

Inconsistency : "or" used `||` but multiple `case` in another location...
```suggestion
            case $presenter->getResponseStatus() instanceof PasswordExpiredResponse:
            case $presenter->getResponseStatus() instanceof UnauthorizedResponse:
            case $presenter->getResponseStatus() instanceof ErrorResponse:
                return View::createRedirect(
                    $this->getBaseUrl() . '/login?authenticationError=' . $presenter->getResponseStatus()->getMessage()
                );
            case $presenter->getResponseStatus() instanceof ErrorAclConditionsResponse:
            case $presenter->getResponseStatus() instanceof ErrorAuthenticationConditionsResponse:
```

File: centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+
+    /**
+     * @var ContactInterface|null
+     */
+    private ?ContactInterface $authenticatedUser = null;
+
+    /**
+     * @var Auth|null
+     */
+    private ?Auth $auth = null;
```
Comment (abuathier) [inline centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php:94]: php-cs-fixer missing
```suggestion
    /** @var Configuration */
    private Configuration $configuration;

    /** @var Centreon */
    private Centreon $legacySession;

    /** @var string */
    private string $username;

    /** @var ContactInterface|null */
    private ?ContactInterface $authenticatedUser = null;

    /** @var Auth|null */
    private ?Auth $auth = null;
```

**Review by sc979 — COMMENTED** [review]
LGTM on automation

**Review by Thebarda — APPROVED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Review by Thebarda — DISMISSED** [review]

**Review by SoldierBoy31 — DISMISSED** [review]

**Review by abuathier — APPROVED** [review]

File: centreon/www/front_src/src/Authentication/SAML/translatedLabels.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  'User ID (login) attribute for Centreon user';
+export const labelEmailAttribute = 'Email attribute';
+export const labelFullNameAttribute = 'Full name attribute';
+export const labelDefineRelationBetweenRolesAndAclAccessGroups =
+  'Define relation between roles and ACL access groups';
+export const labelLogoutFrom = 'Logout from';
+export const labelLogoutUrl = 'Logout URL';
+export const labelCentreonUIOnly = 'Centreon UI only';
+export const labelBothIdentityProviderAndCentreonUI =
+  'Both identity Provider and Centreon UI';
```
Comment (cg-tw) [inline centreon/www/front_src/src/Authentication/SAML/translatedLabels.ts:23]:
```suggestion
  'Both identity provider and Centreon UI';

```

File: centreon/www/front_src/src/Authentication/SAML/translatedLabels.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+export const labelSAMLOnly = 'SAML only';
+export const labelRemoteLoginUrl = 'Remote login URL';
+export const labelEntityIdURL = 'Issuer (Entity ID) URL';
+export const labelCertificate = 'Copy/paste x.509 certificate';
+export const labelUserIdAttribute =
+  'User ID (login) attribute for Centreon user';
+export const labelEmailAttribute = 'Email attribute';
+export const labelFullNameAttribute = 'Full name attribute';
+export const labelDefineRelationBetweenRolesAndAclAccessGroups =
+  'Define relation between roles and ACL access groups';
```
Comment (cg-tw) [inline centreon/www/front_src/src/Authentication/SAML/translatedLabels.ts:18]:
```suggestion
  'Define relations between roles and ACL access groups';

```

**Review by a-launois — DISMISSED** [review]

File: centreon/www/front_src/src/Authentication/shared/translatedLabels.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+export const labelMixed = 'Mixed';
+export const labelEnableAutoImport = 'Enable auto import';
+export const labelContactTemplate = 'Contact template';
+export const labelEnableAutomaticManagement = 'Enable automatic management';
+export const labelGroupsAttributePath = 'Groups attribute path';
+export const labelGroupValue = 'Group value';
+export const labelContactGroup = 'Contact group';
+export const labelDeleteRelation = 'Delete the relation';
+export const labelDefinedTheRelationBetweenGroupsAndContactGroups =
+  'Defined the relation between groups and contact groups';
```
Comment (cg-tw) [inline centreon/www/front_src/src/Authentication/shared/translatedLabels.ts:11]: "Defined" in the past tense?

**Review by abuathier — DISMISSED** [review]

**Review by cg-tw — DISMISSED** [review]

**Review by Thebarda — APPROVED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

**Review by Thebarda — APPROVED** [review]

**Review by abuathier — APPROVED** [review]

**Review by cg-tw — APPROVED** [review]

**Review by sc979 — APPROVED** [review]

**Review by Thebarda — APPROVED** [review]

**Review by SoldierBoy31 — APPROVED** [review]

**Pull request merged**